internal/linker/linker.go

```diff
@@ -0,0 +1,240 @@
+			return nil, err
+		}
+
+		if err := applyPatch(workingDirectory, patch); err != nil {
+			return nil, fmt.Errorf("apply patch for %s failed: %v", fileName, err)
+		}
+	}
+	return mod, nil
+}
+
+func cachePath(goExe string) (string, error) {
+	var cacheDir string
+	if val, ok := os.LookupEnv(garbleCacheDir); ok {
+		cacheDir = val
+	} else {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			panic(fmt.Errorf("cannot retreive user cache directory: %v", err))
+		}
+		cacheDir = userCacheDir
+	}
+
+	cacheDir = filepath.Join(cacheDir, cacheDirName)
+	if err := os.MkdirAll(cacheDir, 0o777); err != nil {
+		return "", err
+	}
+
+	// Note that we only keep one patched and built linker in the cache.
+	// If the user switches between Go versions or garble versions often,
+	// this may result in rebuilds since we don't keep multiple binaries in the cache.
+	// We can consider keeping multiple versions of the binary in our cache in the future,
+	// similar to how GOCACHE works with multiple built versions of the same package.
+	return filepath.Join(cacheDir, "link"+goExe), nil
+}
+
+func getCurrentVersion(goVersion, patchesVer string) string {
+	return goVersion + " " + patchesVer
+}
+
+func checkVersion(linkerPath, goVersion, patchesVer string) (bool, error) {
+	versionPath := linkerPath + versionExt
+	version, err := os.ReadFile(versionPath)
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+
+	return string(version) == getCurrentVersion(goVersion, patchesVer), nil
+}
+
+func writeVersion(linkerPath, goVersion, patchesVer string) error {
+	versionPath := linkerPath + versionExt
+	return os.WriteFile(versionPath, []byte(getCurrentVersion(goVersion, patchesVer)), 0o777)
+}
+
+func compileLinker(workingDirectory string, overlay map[string]string, outputLinkPath string) error {
+	file, err := json.Marshal(&struct{ Replace map[string]string }{overlay})
+	if err != nil {
+		return err
+	}
+	overlayPath := filepath.Join(workingDirectory, "overlay.json")
+	if err := os.WriteFile(overlayPath, file, 0o777); err != nil {
+		return err
+	}
+
+	cmd := exec.Command("go", "build", "-overlay", overlayPath, "-o", outputLinkPath, "cmd/link")
+	// Explicitly setting GOOS and GOARCH variables prevents conflicts during cross-build
+	cmd.Env = append(os.Environ(), "GOOS="+runtime.GOOS, "GOARCH="+runtime.GOARCH)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("compiler compile error: %v\n\n%s", err, string(out))
+	}
+
+	return nil
+}
+
+func PatchLinker(goRoot, goVersion, goExe, tempDirectory string) (string, error) {
+	patchesVer, patches, err := loadLinkerPatches()
+	if err != nil {
+		panic(fmt.Errorf("cannot retrieve linker patches: %v", err))
+	}
+
+	outputLinkPath, err := cachePath(goExe)
+	if err != nil {
+		return "", err
+	}
+
+	mutex := lockedfile.MutexAt(outputLinkPath + ".lock")
+	unlock, err := mutex.Lock()
+	if err != nil {
+		return "", err
+	}
+	defer unlock()
```

Comment:
this is still buggy, I think. you can only release the lock _after the linker has finished running_, per my earlier comment.